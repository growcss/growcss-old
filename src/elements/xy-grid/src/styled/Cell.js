//@flow
import styled from 'styled-components';
import remcalc from 'remcalc';

/**
 * Calculate the percentage size of a cell.
 *
 * @param {String|Number} size
 *
 * @return {String}
 */
const cellSize = (size: string | number): string => {
  if (typeof size === 'string' && size.includes('%')) {
    return size;
  }

  const value = size * 100;

  return value.toString();
};

/**
 * Sets base flex properties for cells.
 *
 * @param {String} size The size of your cell. Accepts `auto`, `shrink` or `grow`.
 *
 * @return {String}
 */
const cellBase = (size: string): string => {
  if (size === 'auto') {
    return 'flex: 1 1 0px;';
  } else if (size === 'shrink') {
    return 'flex: 0 0 auto;';
  } else if (size === 'grow') {
    return 'flex: 1 0 auto;';
  }

  // This is the base style, all others inherit from it
  return 'flex: 0 0 auto;' +
    'min-height: 0px;' +
    'min-width: 0px;';
};

/**
 * Resets a cells width (or height if vertical is true) as well as strips its gutters.
 *
 * @param {Boolean} vertical Set to true to output vertical (height) styles rather than widths.
 *
 * @return {String}
 */
const cellRest = (vertical: boolean): string => {
  const direction: string = vertical === true ? 'height' : 'width';

  return `${direction}: auto;` +
  `max-${direction}: none;`
}

/**
 *  Sets our cell widths or heights depending on gutter type.
 *
 * @param size
 * @param marginGutter
 * @param vertical
 *
 * @return {String}
 */
const cellProperies = (size: string, marginGutter, vertical: boolean): string => {
  const direction: string = vertical ? 'height' : 'width';

  if (vertical === 'full') {
    const val = (marginGutter === 0 ? '100%' : `calc(100% - ${remcalc(marginGutter)})`);

    return `${direction}: ${val};`;
  } else if (vertical === 'auto') {
    return `${direction}: auto;`;
  } else if (vertical === 'shrink') {
    return `${direction}: auto;`;
  }

  const val = marginGutter === 0 ? cellSize(size) : `calc(${cellSize(size)} - ${remcalc(marginGutter)})`;

  return `${direction}: ${val};`;
};

const Cell = styled.div`

`;
