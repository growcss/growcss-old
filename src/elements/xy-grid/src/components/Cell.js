//@flow
import React, { Component } from "react";
import { CellElement } from '../styled/CellElement';

type CellType = {
  children?: any,
  gridColumns: number,
  cellSize?: string,
  gutterType?: string,
  small?: number,
  medium?: number,
  large?: number,
  xlarge?: number,
  xxlarge?: number,
  vertical: boolean
};

export default class Cell extends Component<CellType>
{
  static defaultProps = {
    gridColumns: 12,
    vertical: false,
  };

  render() {
    const {
      children,
      gridColumns,
      vertical,
      cellSize,
      gutterType,
      small,
      medium,
      large,
      xlarge,
      xxlarge,
    } = this.props;

    return (<CellElement
      gridColumns={gridColumns}
      small={small}
      medium={medium}
      large={large}
      xlarge={xlarge}
      xxlarge={xxlarge}
      cellSize={cellSize}
      gutterType={gutterType}
      vertical={vertical}
      >
        {children}
      </CellElement>
    );
  }
}
