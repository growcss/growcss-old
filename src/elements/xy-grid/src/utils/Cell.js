//@flow
import type { GuttersType } from '../types/GuttersType';
import { Gutters as DefaultGutters } from '../components/Gutters';
import { Gutters } from './Gutters';
import { CellBase } from './CellBase';
import { CellProperties } from './CellProperties';

/**
 * Creates a cell for your grid.
 *
 * @param {string}               size           The size of your cell. Can be `full` (default) for 100% width, `auto` to use up available space and `shrink` to use up only required space.
 * @param {boolean}              outputGutter   Whether or not to output gutters.
 * @param {number | GuttersType} gutters        array or number value for gutters.
 * @param {string}               gutterType
 * @param {Array<string>}        gutterPosition The position to apply gutters to. Accepts `top`, `bottom`, `left`, `right` in any combination.
 * @param {string}               breakpoint     The name of the breakpoint size in your gutters array to get the size from.
 * @param {boolean}              vertical       Set to true to output vertical (height) styles rather than widths.
 *
 * @return {string}
 */
export const Cell = (
  size: string = 'full',
  outputGutter: boolean = true,
  gutters: number | GuttersType = DefaultGutters,
  gutterType: string = 'margin',
  gutterPosition: Array<string>  = ['right', 'left'],
  breakpoint: string = 'small',
  vertical: boolean = false,
) => {
  let gutter;

  if (typeof gutters === 'object' && breakpoint in gutters) {
    gutter = gutters[breakpoint];
  } else if (typeof gutters === 'number') {
    gutter = gutters;
  } else {
    throw new Error(`No gutters were found in "${gutters}" for "breakpoint: ${breakpoint}", cell was not generated.`);
  }

  let css = '';

  css += CellBase(size);

  if (gutterType === 'margin') {
    css += CellProperties(size, gutter, vertical);
  } else {
    css += CellProperties(size, 0, vertical);
  }

  if (outputGutter === true) {
    const guttersValue = Gutters(gutter, gutterType, gutterPosition);

    if (Array.isArray(guttersValue)) {
      for (const value: string in guttersValue) {
        css += value;
      }
    } else {
      css += guttersValue;
    }
  }

  return css;
};

