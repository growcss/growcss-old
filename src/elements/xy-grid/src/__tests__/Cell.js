//@flow
import React from 'react';
import Cell from '../components/Cell';
import 'jest-styled-components'

test('cell base layout', () => {
  const wrapper = shallow(<Cell>test</Cell>);

  expect(wrapper).toMatchSnapshot();
  expect(wrapper).toHaveStyleRule('flex', '0 0 auto');
  expect(wrapper).toHaveStyleRule('min-height', '0px');
  expect(wrapper).toHaveStyleRule('min-width', '0px');
  expect(wrapper).toHaveStyleRule('width', '100%');
});

test('cell auto layout', () => {
  const wrapper = shallow(<Cell cellSize='auto'>test</Cell>);

  expect(wrapper).toMatchSnapshot();
  expect(wrapper).toHaveStyleRule('flex', '1 1 0px');
  expect(wrapper).toHaveStyleRule('min-height', '0px');
  expect(wrapper).toHaveStyleRule('min-width', '0px');
  expect(wrapper).toHaveStyleRule('width', 'auto');
});

test('cell shrink layout', () => {
  const wrapper = shallow(<Cell cellSize='shrink'>test</Cell>);

  expect(wrapper).toMatchSnapshot();
  expect(wrapper).toHaveStyleRule('flex', '0 0 auto');
  expect(wrapper).toHaveStyleRule('min-height', '0px');
  expect(wrapper).toHaveStyleRule('min-width', '0px');
  expect(wrapper).toHaveStyleRule('width', 'auto');
});
