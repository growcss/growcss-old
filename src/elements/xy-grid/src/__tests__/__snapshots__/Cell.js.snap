// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cell auto layout 1`] = `
<CellElement
  cellSize="auto"
  gridColumns={12}
  vertical={false}
>
  test
</CellElement>
`;

exports[`cell base layout 1`] = `
<CellElement
  gridColumns={12}
  vertical={false}
>
  test
</CellElement>
`;

exports[`cell shrink layout 1`] = `
<CellElement
  cellSize="shrink"
  gridColumns={12}
  vertical={false}
>
  test
</CellElement>
`;
